### PR TITLE
Compile System.Private.CoreLib with Crossgen2 by default

### PR DESCRIPTION
**Pull request by trylek**
After I deleted the Utf8String-related experimental code from the
runtime repo per JanK's suggestion, I can now enable Crossgen2
compilation for System.Private.CoreLib by default.

Thanks

Tomas

/cc: @dotnet/crossgen-contrib

### CODE REVIEW

**Comment by stephentoub** [comment]
Does this have an impact on code quality, or is the before/after identical?

**Comment by trylek** [comment]
Our measurements so far indicate that Crossgen- and Crossgen2-compiled SPC should be almost identical with regards to runtime performance, working set and size on disk. Having said that, merging this commit will throw a much broader set of tests at the library so I cannot rule out some outlier popping up, that's why this change is intentionally trivial to revert. @mangod9 also suggested that I shoot a heads-up e-mail to the runtime repo committers once this goes in to help identify the cause of any issues that may appear.

**Comment by stephentoub** [comment]
Ok, thanks.

**Review by mangod9 — APPROVED** [review]

**Comment by mangod9** [comment]
There appears to be a build failure on OSX?

**Comment by janvorli** [comment]
The Linux x64 crossgen2 invocation failed. It has failed with an assert:
```
 Process terminated. Assertion failed.
  Please update intrinsic hash table
     at Internal.JitInterface.CorInfoImpl.InitializeIntrinsicHashtable() in /_/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.Intrinsics.cs:line 115
     at Internal.JitInterface.CorInfoImpl..cctor() in /_/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.Intrinsics.cs:line 120
     at Internal.JitInterface.CorInfoImpl.ObjectToHandle(Object obj) in /_/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:line 421
     at Internal.JitInterface.CorInfoImpl.ObjectToHandle(MethodDesc method) in /_/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:line 439
     at Internal.JitInterface.CorInfoImpl.Get_CORINFO_METHOD_INFO(MethodDesc method, MethodIL methodIL, CORINFO_METHOD_INFO* methodInfo) in /_/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:line 455
     at Internal.JitInterface.CorInfoImpl.CompileMethodInternal(IMethodNode methodCodeNodeNeedingCode, MethodIL methodIL) in /_/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.cs:line 147
     at Internal.JitInterface.CorInfoImpl.CompileMethod(MethodWithGCInfo methodCodeNodeNeedingCode) in /_/src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs:line 391
     at ILCompiler.ReadyToRunCodegenCompilation.<ComputeDependencyNodeDependencies>b__26_0(DependencyNodeCore`1 dependency) in /_/src/coreclr/src/tools/aot/ILCompiler.ReadyToRun/Compiler/ReadyToRunCodegenCompilation.cs:line 500
     at System.Threading.Tasks.Parallel.<>c__DisplayClass33_0`2.<ForEachWorker>b__0(Int32 i)
     at System.Threading.Tasks.Parallel.<>c__DisplayClass19_0`1.<ForWorker>b__1(RangeWorker& currentWorker, Int32 timeout, Boolean& replicationDelegateYieldedBeforeCompletion)
     at System.Threading.Tasks.TaskReplicator.Replica`1.ExecuteAction(Boolean& yieldedBeforeCompletion)
     at System.Threading.Tasks.TaskReplicator.Replica.Execute()
     at System.Threading.Tasks.TaskReplicator.Replica.<>c.<.ctor>b__4_0(Object s)
     at System.Threading.Tasks.Task.InnerInvoke()
     at System.Threading.Tasks.Task.<>c.<.cctor>b__277_0(Object obj)
     at System.Threading.ExecutionContext.RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, Object state)
     at System.Threading.Tasks.Task.ExecuteWithThreadLocal(Task& currentTaskSlot, Thread threadPoolThread)
     at System.Threading.Tasks.Task.ExecuteEntryUnsafe(Thread threadPoolThread)
     at System.Threading.Tasks.Task.ExecuteFromThreadPool(Thread threadPoolThread)
     at System.Threading.ThreadPoolWorkQueue.Dispatch()
     at System.Threading._ThreadPoolWaitCallback.PerformWaitCallback()

```

**Comment by janvorli** [comment]
The win x86 and OSX x64 too

**Comment by trylek** [comment]
Got it, @EgorBo's change

https://github.com/dotnet/runtime/commit/819a3cc69e2bf606787a44f9f107ebbbed0fcb11#diff-ff19628c24a151a3f919c197c4785be0c9b10fa61f2e61beb4e4867b676dcc8c

removed two intrinsics without updating the assertion check at

https://github.com/dotnet/runtime/blob/4c102bdf5c68fa5f027b224a23e11c24e385c49b/src/coreclr/src/tools/Common/JitInterface/CorInfoImpl.Intrinsics.cs#L115

I'll send out the fix as a separate PR to keep this switch-over PR trivial to rollback.

**Comment by trylek** [comment]
PR is out: https://github.com/dotnet/runtime/pull/44624

**Comment by AndyAyersMS** [comment]
Before you merge, it would be good for @DrewScoggins to verify we're not seeing any perf lab outage, so if we see any regressions  we have a dense set of runs to help pin down which change is responsible.

**Comment by trylek** [comment]
I've booked vacation from Friday till next Tuesday. Assuming I manage to check in the fix for the assertion check tomorrow morning before I leave, I would be probably inclined to check this in on Saturday or Sunday - this should hopefully also make it easier to identify potential fallout of this change as the rate of check-ins is usually lower over the weekend.

**Comment by trylek** [comment]
Currently lab testing in an instrumented run including the fix for the assertion check: https://github.com/dotnet/runtime/pull/44625

**Comment by AndyAyersMS** [comment]
cc @dotnet/jit-contrib 

Please make sure there's a good writeup somewhere on how to repro and debug crossgen2 errors (particularly what happens if the jit asserts or crashes, and how to pass complus settings), as the first problems a jit developer often sees with new jit code surface during the crossgen of SPC that's done by the build.

We also need to make sure PMI, SPMI, and our automated SPMI collection processes are ready. @sandreenko weren't you looking at this recently?

**Comment by mangod9** [comment]
@AndyAyersMS crossgen2 has been validated with various weekly CI jobs, this change essentially now compiles s.p.corelib.dll by default. Is your concern mainly that there are going to be some JIT failures which might be difficult to debug? Regardless we can certainly get a runbook created for debugging steps.

**Comment by MichalStrehovsky** [comment]
> The Linux x64 crossgen2 invocation failed. It has failed with an assert:

Btw, I still have #43017 out to make the Checked build actually different from a Release build and have asserts fire consistently. I need a signoff to merge, but I guess we need to fix the regression first.

**Comment by DrewScoggins** [comment]
@trylek Could you just comment out the line `pr: none` in https://github.com/dotnet/runtime/blob/master/eng/pipelines/coreclr/perf.yml. This way we can run through the performance tests to make sure that none of them are broken by this change. You can then remove the comment before you check-in.

**Comment by mangod9** [comment]
@DrewScoggins, I have pushed a change to run perf scenarios.

**Comment by mangod9** [comment]
> Thus it's essential that the debugging experience be well documented, eg what command lines are needed, how to specify which jit is being invoked, how to tunnel through the complus settings that are useful when investigating jit issues.

Yeah thats fair. We will work on putting these together.

**Comment by sandreenko** [comment]
SPMI is ready for crossgen2, not sure if we have jitutils support for it to do asm diffs.

However, the debugging experience with crossgen2 was worse than crossgen for me in the past, because:
1. was not clear how to compile as simple as  "crossgen.exe target.dll", instead, you need to know how to create a right ".rsp" file and use it;
2. passing COMPlus_* variables is harder for debugging, you need to pass them as `--codegenopt NgenDump=Repro.Program:Main` in the command line and if I remember correctly some were not working (for example Dump).

**Comment by mangod9** [comment]
@DrewScoggins could you please check the Linux x64 runtime-perf failures, if they are perf related?

**Comment by DrewScoggins** [comment]
Those failures are unrelated.

**Comment by mangod9** [comment]
Ok, thanks for confirming. We will work on putting a runbook together before merging (also need to look at the installer linux-x64 failures if they are related to this change).

**Comment by erozenfeld** [comment]
>> SPMI is ready for crossgen2, not sure if we have jitutils support for it to do asm diffs.

I added support for asm diffs with crossgen2 a year ago in https://github.com/dotnet/jitutils/pull/228 and https://github.com/dotnet/jitutils/pull/234. It may need to be tweaked if crossgen2 command line args have changed since then. Also, we'll probably need to control crossgen2 multithreading.

**Comment by kunalspathak** [comment]
Curious, how do we see the perf results?

**Comment by DrewScoggins** [comment]
These results are not saved, nor are they even applicable. These tests just run the tests on VMs for a single iteration to ensure that they have not been broken by the change, not to investigate regressions or improvements.

**Comment by trylek** [comment]
The OSX x64 failure is known, https://github.com/dotnet/runtime/issues/45088.

**Comment by trylek** [comment]
OK, after one retry the PR run is green now. I have triggered the outerloop run and I intend to merge this in tomorrow European morning (around PST midnight) assuming it reasonably passes and no-one objects. I believe this to be the perfect moment - around Thanksgiving the traffic is pretty low so any potential regression will stand out. As discussed previously, I'll send a heads-up e-mail to the runtime repo committers to watch out for any weird behavior after the merge. Thanks everyone for help with various aspects of this change!

**Comment by mangod9** [comment]
sounds like a good plan. Thanks @trylek

**Review by mangod9 — APPROVED** [review]

**Comment by am11** [comment]
Other than the file size regression, crossgen2 is great! 🙂
```sh
# ubuntu x64
$ ./build.sh -c Release
$ du -h artifacts/bin/coreclr/Linux.x64.Release/System.Private.CoreLib.dll

# before
8.9M	artifacts/bin/coreclr/Linux.x64.Release/System.Private.CoreLib.dll
# after
9.6M	artifacts/bin/coreclr/Linux.x64.Release/System.Private.CoreLib.dll
```

**Comment by trylek** [comment]
Thanks @am11 for pointing that out, I'll take a look. That is surprising as the last time I looked, the CG2 version was actually shorter.

**Comment by trylek** [comment]
OK, so I took a first look using the R2RDump --diff function. I tried this on Windows x64 as there's basically no size difference between the Windows and Linux version (the only actual difference being slightly different calling convention implying subtle changes in some JITted method lengths). The total difference is about 850 KB; almost half of this delta, 380 KB, corresponds to the section <code>DelayLoadMethodCallChunks</code> - in the CG1 version this section only has 33 bytes. @janvorli - am I right to recall it was you who implemented this section? Is it possible that CG1 uses some more compact encoding e.g. by representing all thunks with a single coalesced interval? Supporting this may require putting the thunks into a subsection of its own, implying subtle tweaks to the object writer but the section builder does support this in general, it should be no big deal. I'm looking for the other half that's currently not accounted for by R2RDump for whatever reason.

**Comment by janvorli** [comment]
> . @janvorli - am I right to recall it was you who implemented this section?

No, it was @nattress in #39063

**Pull request merged**
